### PR TITLE
[SIL][PackageCMO] Allow optimizing [serialized_for_pkg] functions 

### PR DESCRIPTION
**Pull request by elsh**
Allow optimizing `[serialized_for_pkg]` functions during SIL inlining, generic/closure specialization, and devirtualization optimization passes.

`SILFunction::canBeInlinedIntoCaller` now exlicitly requires a caller's `SerializedKind_t` arg. `isAnySerialized()` is added as a convenience function that checks if `[serialized]` or `[serialized_for_pkg]`.

Resolves rdar://128704752

### CODE REVIEW

**Comment by elsh** [comment]
@swift-ci test

File: lib/SIL/Verifier/SILVerifier.cpp [hunk 1]
```diff
@@ -158,7 +158,7 @@ namespace {
 template <typename DeclType>
 bool checkResilience(DeclType *D, ModuleDecl *accessingModule,
                      ResilienceExpansion expansion,
-                     bool isSerializedForPackage) {
+                     bool serializedKind) {
```
Comment (aschwaighofer) [inline lib/SIL/Verifier/SILVerifier.cpp:161]: Oversight. This should be `SerializedKind_t serializedKind` not `bool`?

**Comment by aschwaighofer** [comment]
We will also want to update the SwiftCompilerSources, they have instances of both usage of `isSerialized` and `hasValidLinkageForFragileRef`.

```
SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyLoad.swift|284| if !function.isSerialized {
SwiftCompilerSources/Sources/Optimizer/Utilities/OptUtils.swift|552| if parentFunction.isSerialized,                                                                                                            
SwiftCompilerSources/Sources/Optimizer/Utilities/OptUtils.swift|554| !calleeFunction.isSerialized {
SwiftCompilerSources/Sources/SIL/Function.swift|123| public var isSerialized: Bool { bridged.isSerialized(
```

This code for example needs updating to take into account our canInlineInto matrix:
```
extension FullApplySite {                                                       
  var canInline: Bool {                                                         
    // Some checks which are implemented in C++                                 
    if !FullApplySite_canInline(bridged) {                                      
      return false                                                              
    }                                                                           
    // Cannot inline a non-inlinable function it an inlinable function.         
    if parentFunction.isSerialized,                                             
       let calleeFunction = referencedFunction,                                 
       !calleeFunction.isSerialized {                                           
      return false                                                              
    }        
 ```

File: lib/SIL/Verifier/SILVerifier.cpp (hunk 1 above)
Comment (elsh) [reply lib/SIL/Verifier/SILVerifier.cpp:161]: oops, good catch!

**Comment by elsh** [comment]
> We will also want to update the SwiftCompilerSources, they have instances of both usage of `isSerialized` and `hasValidLinkageForFragileRef`.
> 
> ```
> SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyLoad.swift|284| if !function.isSerialized {
> SwiftCompilerSources/Sources/Optimizer/Utilities/OptUtils.swift|552| if parentFunction.isSerialized,                                                                                                            
> SwiftCompilerSources/Sources/Optimizer/Utilities/OptUtils.swift|554| !calleeFunction.isSerialized {
> SwiftCompilerSources/Sources/SIL/Function.swift|123| public var isSerialized: Bool { bridged.isSerialized(
> ```
> 
> This code for example needs updating to take into account our canInlineInto matrix:
> 
> ```
> extension FullApplySite {                                                       
>   var canInline: Bool {                                                         
>     // Some checks which are implemented in C++                                 
>     if !FullApplySite_canInline(bridged) {                                      
>       return false                                                              
>     }                                                                           
>     // Cannot inline a non-inlinable function it an inlinable function.         
>     if parentFunction.isSerialized,                                             
>        let calleeFunction = referencedFunction,                                 
>        !calleeFunction.isSerialized {                                           
>       return false                                                              
>     }        
> ```

I thought bridged functions should just be either serialized or not_serialized, i.e. package kind doesn't apply here. Or am I missing something? Also a specializer that calls hasValidLinkageForRef in this dir is for a feature that's not supported.

**Comment by elsh** [comment]
@swift-ci test

**Comment by elsh** [comment]
@swift-ci test macOS

**Comment by aschwaighofer** [comment]
The added changes since my last review look good to me. We also need a change in SwiftCompilerSources/Sources/Optimizer/Utilities/OptUtils.swift to take serializedForPackage into account:

```
     // Cannot inline a non-inlinable function it an inlinable function.         
    if parentFunction.isSerialized,                                             
       let calleeFunction = referencedFunction,                                 
       !calleeFunction.isSerialized {                                           
      return false                                                              
    } 
 ```
```

**Comment by elsh** [comment]
@swift-ci test

**Comment by elsh** [comment]
> The functional part makes sense to me. Can we add tests to ensure that these optimization passes are actually active for serialized_for_pkg functions?

More tests including this are in the upcoming PR.

**Pull request merged**